Reduction of unnecessary downloads from GPU

alphaville opened the pull request:
## Main Changes

- Extra checks only if `GPUTILS_DEBUG_MODE` is defined
- methods do not return a status code now
- status code accessible with a getter (docs written)

## Associated Issues

None

alphaville committed:
Reduction of unnecessary downloads from GPU

About:
- Extra checks only if GPUTILS_DEBUG_MODE is defined
- methods do not return a status code now
- status code accessible with a getter (docs written)

alphaville committed:
Introduce class IStatus

About:
- IStatus is subclassed by all classes that need a device-based status code

alphaville committed:
Define IStatus constructor

alphaville committed:
Merge main and fix conflicts in CHANGELOG

ruairimoran commented on include/tensor.cuh:
After line 2035 (in updated code), we need `TEMPLATE_CONSTRAINT_REQUIRES_FPX` to remove `Requires clause differs in template redeclaration` warning

ruairimoran commented on include/tensor.cuh:
```diff
@@ -1225,8 +1254,14 @@ private:
+
 public:
 
+    int getStatus() {
```
should this not be `statusCode()`

ruairimoran commented on include/tensor.cuh:
```diff
@@ -1225,8 +1254,14 @@ private:
 
+    int getStatus() {
+        // redefinition of getStatus
+        return m_svdStatus;
```
why not have `m_svdStatus = std::max(m_svdStatus, (*m_info)(0, 0, 0));` here, so you don't need the `#ifdef GPUTILS_DEBUG_MODE` blocks ? 

Why even have m_svdStatus ? We can redefine `statusCode()` with `return (*m_info)(0, 0, 0)`. You will want to call `statusCode()` after every method when debugging anyway, so you catch the first instance of a problem

alphaville replied:
I've changed the code slightly:

1. In `CholeskyBatchFactoriser` we need to allocate more memory for the device info; this is required by cuBLAS; see API docs of  [`potrsbatched`](https://docs.nvidia.com/cuda/cusolver/index.html#cusolverdn-t-potrsbatched) 
2. In `Svd` we perform multiple calls of `cusolverDnDgesvd` in a for loop and we result `m_info`. If the maximum of all these `m_info` is zero, all is good. 

Anyway, let's talk about this on Teams.

ruairimoran reviewed (commented):
some minor comments. Also, there appears to be a memory leak

alphaville committed:
Fix memory issue (satanizer complained)

alphaville committed:
try again

alphaville committed:
Fix memory issues

About:
- IStatus allocates appopriate memory
- Rename getStatus() to statusCode()

alphaville committed:
Use TEMPLATE_CONSTRAINT_REQUIRES_FPX in Givens

alphaville committed:
testing something quick (memory)

ruairimoran reviewed (approved):
lgtm

alphaville committed:
Change statusCode to info (returns tensor ref)

alphaville committed:
update CHANGELOG

alphaville committed:
fix issue in main

alphaville committed:
changelog: update release date

alphaville committed:
final touch: info in SVD
